### PR TITLE
yuzu-ea: 3702 -> 3783, yuzu-mainline: 1475 -> 1513 

### PR DESCRIPTION
**Pull request by leiserfg**
###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by leiserfg** [comment]
This is not the latest version, but we can't bump it to latest without bumping Vulkan first.

**Comment by leiserfg** [comment]
@SuperSandro2000 @PedroHLC @Ralith can you please give some feedback?

File: pkgs/applications/emulators/yuzu/generic.nix
Comment (Ralith) [inline pkgs/applications/emulators/yuzu/generic.nix:116]: This is a submodule checkout of the upstream Vulkan-Headers repo, right? If `YUZU_USE_EXTERNAL_VULKAN_HEADERS=ON` is set, why do we need to mess with it at all?

**Review by Ralith — CHANGES_REQUESTED** [review]

File: pkgs/applications/emulators/yuzu/generic.nix
Comment (PedroHLC) [reply pkgs/applications/emulators/yuzu/generic.nix:116]: Humm, the way vulkan works, one should be able to connect an application with an outdated loader/driver that does not provide a specific extension or api-level. But I imagine in compilation-time the same does not apply, since the symbol will be MIA gcc should fail.

Comment (Ralith) [reply pkgs/applications/emulators/yuzu/generic.nix:116]: I'm not sure what that has to do with my comment, but it's wrong in any case. Vulkan extensions/optional features are loaded dynamically; they're not statically exported symbols at all. The linker isn't involved.

Comment (PedroHLC) [reply pkgs/applications/emulators/yuzu/generic.nix:116]: I meant to comment the same line, and by the compile-time I meant using the symbols defined in the vulkan's headers in a newer version but not older one (that will eventually happen considering our use-case)

Comment (Ralith) [reply pkgs/applications/emulators/yuzu/generic.nix:116]: Updates to the Vulkan headers do not introduce new statically exported symbols. That would make it impossible to develop software that gracefully degrades on older systems.

Comment (PedroHLC) [reply pkgs/applications/emulators/yuzu/generic.nix:116]: Ok, that sounds ideal.

**Comment by K900** [comment]
I'd rather update the Vulkan stuff in nixpkgs, or at least override our derivation and use it as an input. Relying on vendored stuff is not great.

File: pkgs/applications/emulators/yuzu/generic.nix
Comment (leiserfg) [reply pkgs/applications/emulators/yuzu/generic.nix:116]: > why do we need to mess with it at all?

When cmake loads Vulkan that module is already defined, if I don't comment that line cmake complaints because the module is being redefined.

**Comment by leiserfg** [comment]
> I'd rather update the Vulkan stuff in nixpkgs

We decided not to keep updating Vulkan to match yuzu's requirements because they use non-stable versiona of the Vulkan components.

**Comment by K900** [comment]
That's not really the case. The version they're using _is_ stable, there are just some release cadence issues that make updating the rest of the Vulkan SDK difficult sometimes.

**Comment by leiserfg** [comment]
> That's not really the case. The version they're using _is_ stable, there are just some release cadence issues that make updating the rest of the Vulkan SDK difficult sometimes.

Some context:
https://github.com/NixOS/nixpkgs/issues/243780#issuecomment-1648627379_

File: pkgs/applications/emulators/yuzu/generic.nix
Comment (leiserfg) [reply pkgs/applications/emulators/yuzu/generic.nix:116]: Originally I tried only setting the cmake flag, but it was not enough. Another option could be overriding the checked version, but that will require some regex mumbo-jumbo.

**Comment by PedroHLC** [comment]
On the Vulkan-nixpkgs-side, I previously proposed we “versioning” it, in at least `_sdk` and `_latest`: https://github.com/NixOS/nixpkgs/pull/229897. It's easier to update every other package in nixpkgs (one goes through staging, the other directly to master), and we have a fallback in case the latest is borked.

I have a simpler solution in chaotic's nyx nowadays: [vulkan-versioned](https://github.com/chaotic-cx/nyx/tree/main/pkgs/vulkan-versioned). But I still would prefer all of them in one directory in nixpkgs to help bumping.

---

On Yuzu-EA-side, we should update Yuzu in separated PRs, always. Yuzu non-ea still depends on 1.3.256, and we can safely update. Meanwhile, the EA **it's updated multiple times during the same day**, includes **non-merged poorly reviewed PRs**. Occasionally, EA doesn't even work, e.g., in 3770 they pushed a requirement on QRegExpValidator which doesn't exist in Qt6 and trashed that build (and two tags following it).

**Comment by leiserfg** [comment]
IMHO as the dependency on latest commits of Vulkan is not real we shouldn't keep having to update it just for the sake of yuzu. Yuzu builds and runs with the stable SDK (I tested the EA locally before pushing the pr), so it's less work if we just update the SDK when there is a new stable (blessed) release.

**Review by PedroHLC — APPROVED** [review]
> I tested the EA locally before pushing the pr

And someone should always do on future bumps, so I have to agree with the changes, and diff LGTM.

(But the mainline one could have been updated earlier on if we did a separated PR.)

File: pkgs/applications/emulators/yuzu/generic.nix
Comment (SuperSandro2000) [inline pkgs/applications/emulators/yuzu/generic.nix:115]:
```suggestion
    substituteInPlace externals/Vulkan-Headers/CMakeLists.txt \
```

**Comment by leiserfg** [comment]
> (But the mainline one could have been updated earlier on if we did a separate PR.)

Agree, but the script to update bumps both versions, so I just did both, and I think that's what happens most of the time.

**Comment by misuzu** [comment]
Last commit should probably be squashed with the first one

File: pkgs/applications/emulators/yuzu/generic.nix
Comment (Ralith) [reply pkgs/applications/emulators/yuzu/generic.nix:116]: > When cmake loads Vulkan that module is already defined, if I don't comment that line cmake complaints because the module is being redefined.

Hmm, it would be better to patch out their loading cmake modules from the vendored `externals/Vulkan-Headers` entirely.

**Comment by leiserfg** [comment]
> Last commit should probably be squashed with the first one

I think it's better to squash when there is nothing else to fix to keep the changes matching the commentaries here until then.

File: pkgs/applications/emulators/yuzu/generic.nix
Comment (leiserfg) [reply pkgs/applications/emulators/yuzu/generic.nix:116]: It's fine if I do:
```bash
    substituteInPlace     externals/CMakeLists.txt \
    --replace "add_subdirectory(Vulkan-Headers)" ""
```
instead ?

Comment (leiserfg) [reply pkgs/applications/emulators/yuzu/generic.nix:116]: That also works.

**Comment by leiserfg** [comment]
BTW, there is a new mainline version of yuzu out already (and like 10 -ea) should I update them in this PR?

**Comment by K900** [comment]
We should be able to use system Vulkan headers after https://github.com/NixOS/nixpkgs/pull/245954 (hopefully) lands.

**Comment by leiserfg** [comment]
We already do it without that PR.

File: pkgs/applications/emulators/yuzu/generic.nix
Comment (Ralith) [reply pkgs/applications/emulators/yuzu/generic.nix]: LGTM, thanks!

**Review by Ralith — APPROVED** [review]
Approving Vulkan-related tweaks; no opinion on yuzu itself.

**Comment by leiserfg** [comment]
I see the other maintainers are inactive in github, is there any hope of getting this merged?

**Comment by leiserfg** [comment]
> If we're patching it out anyway, why not just downgrade the requested version [here](https://github.com/yuzu-emu/yuzu/blob/master/CMakeLists.txt#L317)?

I thought about doing it before going this route and ended up with this to avoid having to add regexes / change this by hand every time they change it upstream.

**Comment by leiserfg** [comment]
> I'd actually prefer a patch here, specifically because it's going to fail in a loud and predictable way next time upstream raises the required version.

If the changes upstream require a new version (for real) it won't compile.

**Comment by K900** [comment]
Yes, with a confusing failure mode that doesn't clearly indicate the root cause of the issue.

**Comment by leiserfg** [comment]
Are you ok with just skipping the version check (using a hardcoded version)? That will also fail if the version mismatches, I can add a comment to clarify the intention and what to do after.

**Comment by K900** [comment]
What do you mean by "using a hardcoded version"?

**Comment by leiserfg** [comment]
I mean:
```sh
  substituteInPlace    CMakeLists.txt \
    --replace "find_package(Vulkan 1.3.256 REQUIRED)" ""
```

**Comment by K900** [comment]
Just removing the line won't work, we need to replace it with a lower version (or an unversioned specification).

**Comment by K900** [comment]
Also, `substituteInPlace` won't fail if there are no matches in the source file - it will print a warning, but it will not fail the build.

**Comment by leiserfg** [comment]
> Just removing the line won't work, we need to replace it with a lower version (or an unversioned specification).

You are right, I forgot it was a find_package not a check, but same idea, if that works for you I can try to do it (and see how can I make it fail if not substitution happens).

**Comment by K900** [comment]
Honestly, I still think the easiest way is to just write a patch. All we really need is 

```patch
-find_package(Vulkan 1.3.256 REQUIRED)
+find_package(Vulkan 1.3.250 REQUIRED)
```

And it's going to fail cleanly, it's going to be very evident what it does, and it's going to be very evident why it's there.

**Comment by Ralith** [comment]
Pretty awkward that this now once again pulls in the vendored Vulkan-Headers' cmake scripts, and perhaps even uses them exclusively. Shouldn't cause any problems, though, so w/e, hopefully upstream will clean it up someday.

**Comment by K900** [comment]
@Ralith, this is another reason the patch is a better option, actually.

@leiserfg, could you update to the latest version, squash all the commits and then maybe add a comment to the patch explaining why it was added?

**Review by K900 — APPROVED** [review]
Would prefer a comment on the patch, but no big deal. Will test a bit more tomorrow.

**Comment by leiserfg** [comment]
I added the comment in the patch itself, it will be easier to find in the future.

**Comment by K900** [comment]
Seems good here. Thanks a lot for investigating this!

**Pull request merged**
